### ISSUE TITLE
Release only if all builds ended successfully (wuth Travis build stages)

### ISSUE THREAD

**Issue by mockitoguy** [mockito/mockito#1101]
I noticed today that Travis CI publishes new release even though the Java8 jobs have not completed yet. This seems odd. I would assume that 'after_success' would wait for all matrix jobs to complete first. @szpak - do you know why it works like that?

**Comment by TimvdLippe** [comment]
Seems like an upstream issue at https://github.com/travis-ci/travis-ci/issues/929 A solution linked there is https://github.com/dmakhno/travis_after_all as well as the official Travis solution at https://docs.travis-ci.com/user/build-stages which has been in open beta 3 weeks ago: https://github.com/travis-ci/beta-features/issues/11

**Comment by szczepiq** [comment]
Thanks a lot for investigation. I suggest that we use the Beta feature of Travis. What do you guys think? If you are ok with that let's update the description of this ticket and mark it 'please contribute' ;)

**Comment by TimvdLippe** [comment]
Yes lets do that

**Comment by szpak** [comment]
@szczepiq No, it doesn't. As Tim already mentioned Travis (at last) started providing some solution for that. I gave it a first try some time ago, but there were multiple issues and Travis' team is working on it (see very long discussion and reports - https://github.com/travis-ci/beta-features/issues/11). I plan to give it the second try in one of my projects in the foreseeable future, so there is a chance to backport that solution to Mockito later on (if stable enough...).

Nevertheless, guys, feel free to do it on you own earlier :).

**Comment by szczepiq** [comment]
@szpak, thank you for useful information and warning :) Perhaps we should switch to circle ci or something else? :)

**Comment by szpak** [comment]
The issues are just with the net beta feature. Not with the Travis in general. It's good to make it available after all those years :). Nevertheless, @szczepiq you can do some PoC and comparison between lightweight CI/CD servers/services. Maybe there is something better :).

**Comment by szczepiq** [comment]
>Nevertheless, @szczepiq you can do some PoC and comparison between lightweight CI/CD servers/services. Maybe there is something better :).

Good idea! In your experience, Travis CI is currently the best option for open source?

**Comment by szpak** [comment]
On a daily basis (for in-house development of µservices and other stuff) I use Jenkins powered by job-dsl (and a custom framework on top of it) to provide pipelines (and jobs) managed in code. It works very well. Unfortunately, although CloudBees offers a hosted Jenkins version (which is in general free for FOSS projects) it has its limitation (such as blacklisted Gradle plugins - mostly due to security issues). Keeping an own instance of Jenkins for FOSS projects behind VPN would be also problematic in the context of external contributors. Therefore, I use Travis for open source stuff which was the best when I gave also some other solutions a try 2 years ago (or so) and which is de facto a standard for GitHub (however, it isn't fully a FOSS software - although much more than GitHub itself :( ). I don't know how Circle CI looks like nowadays. For sure GitLab CI is a very promising tool, but it is rather limited to projects hosted on GitLab (which is still less popular for FOSS projects than GH).

To sum it up - it would be good to see an objective and uptodate comparison done in a real project in mind (such as Mockito).

Nevertheless, I don't believe in a Mockito migration in the foreseeable future - shipkit would need to be extended to support it which IHMO should not a feature of 1.0 (as Travis seems to be much more popular in FOSS (GitHub) projects than any other solution).

**Comment by TimvdLippe** [comment]
It seems like we can now try this out on Travis: https://github.com/travis-ci/beta-features/issues/11#issuecomment-329188863

**Comment by slawekjaranowski** [comment]
Is it still actual?
Travis was removed from build https://github.com/mockito/mockito/commit/5bf62a389c242364ec87d200ae845aba0ea5a721

**Comment by mockitoguy** [comment]
We can close it.
